sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -659,6 +669,40 @@ class Analyzer(
         }
         Sort(newOrders, global, child)
 
+      // Replace the index with the corresponding expression in aggregateExpressions. The index is
+      // a 1-base position of aggregateExpressions, which is output columns (select expression)
+      case a @ Aggregate(groups, aggs, child)
+          if conf.groupByOrdinal && child.resolved && aggs.forall(_.resolved) &&
+            groups.exists(IntegerIndex.unapply(_).nonEmpty) =>
+        val newGroups = groups.map {
+          case IntegerIndex(index) if index > 0 && index <= aggs.size =>
+            aggs(index - 1) match {
+              case e if ResolveAggregateFunctions.containsAggregate(e) =>
+                throw new UnresolvedException(a,
+                  s"Group by position: the '$index'th column in the select contains an " +
+                  s"aggregate function: ${e.sql}. Aggregate functions are not allowed in GROUP BY")
+              // Group by clause is unable to use the alias defined in aggregateExpressions
```

Comment:
This part is used to handle the case like:
```SQL
SELECT a, (b + 2) as c, count(2) FROM testData2 GROUP BY a, 2
```
If we do not have this case, I think our Spark SQL is unable to handle the following converted plan:
```SQL
SELECT a, (b + 2) as c, count(2) FROM testData2 GROUP BY a, c
```

In addition, our parser does not support the following query:
```SQL
SELECT 1, sum(b) FROM testData2 GROUP BY (b + 1) as c
```